Improvements to zoom dropdown behaviour

eldang opened the pull request:
- [x] add switch to house|senate districts link under the zoomto dropdown (show both options)
- [x] can we turn the "State House|Senate Districts" item at the top of the zoom-to- dropdown into a zoom back out to the whole state option?
- [x] can I make zoom actions always update the URL?
- [x] wording changes requested by Max 12/20
- [x] extend timing for populating the stats card yet further to improve iPad reliability
- [x] highlight the zoomed-to district

eldang committed:
Added link to switch between Senate & House districts

eldang committed:
Made more info text only mention the relevant district type

eldang committed:
Zooming to a district updates the URL

eldang committed:
Now also updates the window title

eldang commented on index.html:
```diff
@@ -225,6 +225,29 @@
 
+			function updateURL(district=null) {
+				newURL = window.location.pathname;
+				newTitle = 'Raise Your Hand Texas programs'
```
Note that `newTitle` will overwrite the current window title, after having some relevant context appended to it.  So this is where to edit if you want the title changed.

eldang replied:
Note that `newTitle` will overwrite the current window title, after having some relevant context appended to it.  So this is where to edit if you want the title changed.

eldang commented on css/ryht-legislative-style.css:
```diff
@@ -352,16 +352,32 @@ p.moreinfo  {
+}
 
-  p.programs  {
+/* I'm starting this as a copy of .moreinfo, with just the top & bottom padding removed */
```
I added a class for the new UI piece that lets the viewer switch between House and Senate districts.  Styling is currently based on `p.moreinfo` but if you want to tweak it, here's where to look.

coregis replied:
This is good, the styling is a nice bit of symmetry with the bottom of the flyout

eldang committed:
Auto-rewrites all URLs to what it thinks is going on

to avoid ambiguity with malformed URL parameters

eldang committed:
Trimmed unused variable

eldang committed:
Forces page reload whenever navigating history

eldang committed:
window title improvements

eldang commented:
One detail I'm not satisfied with: when the user clicks the back button, it goes back appropriately, but somehow the history to go _forward_ is lost.

eldang committed:
Top option in the zoomto dropdown now zooms back out to state level

eldang committed:
Zooming out through the dropdown now also clears layer filters

eldang commented:
I'm done with this for today (Wednesday), and it's in a safe state to merge.  I will come back to highlighting the selected district tomorrow, whether this PR's been merged yet or not.

eldang committed:
syntax consistency

coregis commented:
Just had a quick look at this branch and I am impressed! Sorry for the radio silence lately, I am heading out to give a 4 hour workshop at TWS this afternoon and will give this my undivided attention starting first thing tomorrow morning.

eldang commented:
No worries!  I figure you're busy enough that I should avoid getting bottlenecked waiting for response from you this week....  I'm hoping to get the last item on the list done today; I'll keep you posted.

eldang committed:
Wording changes per client's request

eldang committed:
Extended the time over which the stats box gets updated

eldang committed:
Highlights zoomed-to district!

eldang commented:
I have the mechanics of highlighting the zoomed-to district working, but I'm not thrilled with the aesthetics.  Working on improvements now....

eldang committed:
Nicer looking active district emphasis

eldang commented on index.html:
```diff
@@ -132,6 +133,9 @@
 						}
 					}
 				}
+				map.setLayoutProperty(sourceID + '-poly', 'visibility', 'none');
```
The way I've made deemphasising the other districts work is that the House or Senate districts polygon layer is now effectively a mask.  There's a potentially confusing sequence of events:

1. On first load, it's technically `visible` but has the 0 alpha value that is set in the `addVectorLayer()` call that first loads it.  This is necessary because it needs to be visible _to mapbox-gl.js_ for my code to be able to find each district's bounding box to set the zoom control up, and leaving it as fully transparent prevents that from being visible _to the user_.
2. Once the `populateZoomControl()` function has finished stepping through all the districts, it sets the polygon layer's `visibility` property to `none`, and then the next two lines set up the colour for the mask - currently a translucent light grey.  This is now safe to do because the layer isn't visible.
3. Then when we zoom to a district, the `zoomToPolygon()` function makes this mask layer visible again, but filters out the one district we're interested in.  That way everything else is behind the translucent mask.

The main practical consequence of this is that if you want to change the appearance of the mask, you should do that by editing the `map.setPaintProperty()` calls here, _not_ the properties with which it's initially loaded, because those will get overridden by whatever's specified here.

coregis replied:
That's pretty slick! And pretty tricky. Thanks for the detailed explanation of what's happening, I have a feeling it will come in useful down the road.

eldang replied:
Indeed!  I'm patting myself on the back for this because it's going to save me reinventing the wheel to add this to the pre-K map.  I probably won't have time to finish it today, though.

eldang committed:
Added comment for future maintenance

eldang committed:
Fixed some awkward stats box bugs

Because updating the stats box involves setting some delays to have it try a few times, it was possible to get weird behaviour if someone jumped between items in the zoom to districts list too quickly.  Specifically, there could be duelling updates between the previous and current district.  This commit fixes that.

eldang commented:
What I have learned about the stats box is that it can only ever count layers which are visible.  This causes a problem for the zoom back out to the whole state, because that action doesn't force every layer to be visible.  This means that if the user turned any layers off and then zooms back out to the whole state, they'll get erroneous zeroes in the stats box.  I see two possible responses to this:

A) We tell RYHT that we can't make the stats box count the whole state.  Drawback: they want it to.
B) We make the zoom out to the whole district also activate all layers.  Drawback: visual crowding.

I'm not sure which is a better idea, but they're both technically achievable.

eldang committed:
Added zoom out link below the district selector

eldang commented on index.html:
```diff
@@ -421,6 +421,8 @@
 
 				<!-- House/Senate districts switcher -->
 
+				<p onClick="zoomToPolygon('', '-108,25,-88,37,0', '')" class="zoomoutlink">Show whole state</p>
```
If you decide against including the zoom out link, you just have to prune this one line.  And if you want to style it differently, `zoomoutlink` is the CSS class to edit.

coregis replied:
Thanks for presenting these options. I think we defer to RYHT's preference on this one after they've had a chance to play with v2.0.
